File: app/src/main/java/com/vikingsen/cheesedemo/model/webservice/WebServiceModule.java
package com.vikingsen.cheesedemo.model.webservice;


import android.os.Build;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.module.SimpleModule;
import com.vikingsen.cheesedemo.BuildConfig;
import com.vikingsen.cheesedemo.model.prefs.Prefs;
import com.vikingsen.cheesedemo.model.webservice.converter.LocalDateTimeDeserializer;
import com.vikingsen.cheesedemo.model.webservice.converter.LocalDateTimeSerializer;

import org.threeten.bp.LocalDateTime;

import javax.inject.Singleton;

import dagger.Module;
import dagger.Provides;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.logging.HttpLoggingInterceptor;
import retrofit2.Retrofit;
import retrofit2.converter.jackson.JacksonConverterFactory;

@Module
public class WebServiceModule {

    private static final String USER_AGENT = BuildConfig.USER_AGENT_APP_NAME + " " +
                                             BuildConfig.VERSION_NAME + " / Android " +
                                             Build.VERSION.RELEASE + " " +
                                             Build.VERSION.INCREMENTAL + " / " +
                                             Build.MANUFACTURER + " " +
                                             Build.MODEL;

    @Provides
    @Singleton
    HttpLoggingInterceptor provideHttpLoggingInterceptor(Prefs prefs) {
        return new HttpLoggingInterceptor().setLevel(prefs.getHttpLoggingLevel());
    }

    @Provides
    @Singleton
    OkHttpClient provideOkHttpClient() {
        OkHttpClient.Builder builder = new OkHttpClient.Builder();
        builder.addNetworkInterceptor(chain -> {
            Request.Builder requestBuilder = chain.request().newBuilder();
            requestBuilder.addHeader("http.useragent", USER_AGENT);
            return chain.proceed(requestBuilder.build());
        });
        return builder.build();
    }

    @Provides
    @Singleton
    ObjectMapper provideObjectMapper() {
        SimpleModule module = new SimpleModule();
        module.addSerializer(new LocalDateTimeSerializer());
        module.addDeserializer(LocalDateTime.class, new LocalDateTimeDeserializer());

        ObjectMapper objectMapper = new ObjectMapper();
        objectMapper.registerModule(module);

        return objectMapper;
    }

    @Provides
    @Singleton
    CheeseService provideCheeseSerivce(OkHttpClient okHttpClient, HttpLoggingInterceptor httpLoggingInterceptor, ObjectMapper objectMapper) {
        OkHttpClient client = okHttpClient.newBuilder().addNetworkInterceptor(httpLoggingInterceptor).build();
        JacksonConverterFactory converterFactory = JacksonConverterFactory.create(objectMapper);

        Retrofit retrofit = new Retrofit.Builder()
                .baseUrl(BuildConfig.SERVICE_BASE_URL)
                .client(client)
                .addConverterFactory(converterFactory)
                .build();

        return retrofit.create(CheeseService.class);
    }
}

File: app/src/main/java/com/vikingsen/cheesedemo/model/webservice/WebServiceModule.kt
package com.vikingsen.cheesedemo.model.webservice


import android.os.Build
import com.google.gson.Gson
import com.google.gson.GsonBuilder
import com.vikingsen.cheesedemo.BuildConfig
import com.vikingsen.cheesedemo.model.prefs.Prefs
import com.vikingsen.cheesedemo.model.webservice.converter.OffsetDateTimeAdapter
import dagger.Module
import dagger.Provides
import okhttp3.OkHttpClient
import okhttp3.logging.HttpLoggingInterceptor
import org.threeten.bp.OffsetDateTime
import retrofit2.Retrofit
import retrofit2.converter.gson.GsonConverterFactory
import javax.inject.Singleton

@Suppress("unused")
@Module
class WebServiceModule {

    @Provides
    @Singleton
    internal fun provideHttpLoggingInterceptor(prefs: Prefs): HttpLoggingInterceptor =
        HttpLoggingInterceptor().setLevel(prefs.httpLoggingLevel)

    @Provides
    @Singleton
    internal fun provideOkHttpClient(): OkHttpClient {
        val builder = OkHttpClient.Builder()
        builder.addNetworkInterceptor { chain ->
            val requestBuilder = chain.request().newBuilder()
            requestBuilder.addHeader("http.useragent", USER_AGENT)
            chain.proceed(requestBuilder.build())
        }
        return builder.build()
    }

    @Provides
    @Singleton
    internal fun provideGson(): Gson {
        return GsonBuilder()
            .registerTypeAdapter(OffsetDateTime::class.java, OffsetDateTimeAdapter())
            .create()
    }

    @Provides
    @Singleton
    internal fun provideCheeseService(okHttpClient: OkHttpClient, httpLoggingInterceptor: HttpLoggingInterceptor, gson: Gson): CheeseService {
        val client = okHttpClient.newBuilder().addNetworkInterceptor(httpLoggingInterceptor).build()
        val converterFactory = GsonConverterFactory.create(gson)

        val retrofit = Retrofit.Builder()
                .baseUrl(BuildConfig.SERVICE_BASE_URL)
                .client(client)
                .addConverterFactory(converterFactory)
                .build()

        return retrofit.create(CheeseService::class.java)
    }

    companion object {
        private val USER_AGENT = "${BuildConfig.USER_AGENT_APP_NAME} ${BuildConfig.VERSION_NAME} / " +
                "Android ${Build.VERSION.RELEASE} ${Build.VERSION.INCREMENTAL} / " +
                "${Build.MANUFACTURER} ${Build.MODEL}"
    }
}
